src/workerd/api/r2-rpc.c++

```diff
@@ -162,9 +163,17 @@ kj::Promise<R2Result> doR2HTTPPutRequest(jsg::Lock& js, kj::Own<kj::HttpClient>
         expectedBodySize = text.value.size();
         KJ_REQUIRE(streamSize == nullptr);
       }
-      KJ_CASE_ONEOF(data, kj::Array<kj::byte>) {
+      KJ_CASE_ONEOF(data, jsg::BufferSource) {
         expectedBodySize = data.size();
         KJ_REQUIRE(streamSize == nullptr);
+        // We want to either detach or copy the given ArrayBuffer/TypedArray
+        // in order to prevent the possibility of post-call modification.
+        if (featureFlags.getDetachArrayBufferOnPut()) {
+          maybeBackingStore = data.detach(js);
```

Comment:
Is it possible for the JS code to modify this buffer before this code is reached? Or is it impossible for the c++ code to be pre-empted between the function call & this code?

Reply:
As long as this function is called synchronously, JS can't modify the buffer before this code is reached.